### DISCUSSION TITLE
[RESOLVED] Warning if variable is re-assigned?

### DISCUSSION THREAD

**Discussion by richb-hanover** [PRQL/prql#1159] [General]
Currently, PRQL sets a variable to its latest assignment, losing/ignoring any earlier value. For example:

```
select [
a = 1,
a = 2,
]
```
yields 
```
SELECT
  2 AS a
```
It seems to me that this is almost always an error - either a simple typo or a deeper problem - and should be flagged as an error.

Is there ever a case where re-assignment could be useful? Thanks.

**Comment by max-sixty** [comment]
I'm less persuaded this is almost always an error. e.g. say we want to replace nulls with zero:

```
from employees
derive salary = (salary ?? 0)
```

Almost all programming languages allow reassignments andor mutations...

**Comment by richb-hanover** [comment]
> Almost all programming languages allow reassignments and/or mutations...

Yes, but the mistake I'm trying to guard against is mistyping a complicated column name (I have plenty of similar ones in my application: _SDAppTotal2021_ and _SDAssTotal2021_) and having it disappear from the resulting table because it got reassigned...

That said, your example is a compelling use case. Furthermore, it "does the right thing" if extended like this:

```
from employees
derive [
  salary = (salary ?? 0),
  salary = salary-7,
]
```
generates:

```
SELECT
  employees.*,
  COALESCE(salary, 0) - 7 AS salary
FROM
  employees
```
It seems there's tension generated by the dual purpose of `variable = expression`:
* create a column in the resulting table
* give a name to that column to use in subsequent calculations.

I was initially thinking about the way Javascript makes a distinction between `const` and `let` in declarations:

```
const a = 1 // compiler can detect re-assignment to 'a' as an error
let b = 1   // re-assignment is allowed
```

I wonder if it's sufficient to require re-assignment to occur on one line: e.g., ` salary = (salary ?? 0) - 7`, so that the appearance of another `salary = ...` could trigger a warning/error...

**Reply by max-sixty** [reply]
There was some discussion on that in https://github.com/prql/prql/issues/6. 

In retrospect, I think we came to the correct decision given the current model's simplicity. Though I can empathize with the stricter model's approach.

**Comment by richb-hanover** [comment]
Thanks for that link to the previous discussion. I'll just have to be careful when typing variable names. (And why wouldn't I want to do this? :-)
